pyi.py

```diff
@@ -671,6 +671,53 @@ def _analyse_union(members: Sequence[ast.expr]) -> UnionAnalysis:
     )
 
 
+class TypingLiteralAnalysis(NamedTuple):
+    members_by_dump: defaultdict[str, list[ast.expr]]
+    members_without_none: list[ast.expr]
+    none_members: list[ast.expr]
+    contains_only_none: bool
+
+
+def _analyse_typing_Literal(node: ast.Subscript) -> TypingLiteralAnalysis:
+    """Return a tuple providing analysis of a `typing.Literal` slice.
+
+    >>> source = 'Literal[True, None, True, None, False]'
+    >>> literal = _ast_node_for(source)
+    >>> analysis = _analyse_typing_Literal(literal)
+    >>> len(analysis.members_by_dump["Constant(value=True)"])
+    2
+    >>> analysis.contains_only_none
+    False
+    >>> len(analysis.none_members)
+    2
+    >>> unparse(ast.Tuple(analysis.members_without_none))
+    '(True, True, False)'
+    """
```

Comment:
There's two things causing this doctest to fail on py38:

### (1) The `ast.dump()` of `Constant` nodes changed in py39

On py38:

```pycon
Python 3.8.16 (default, Mar  2 2023, 03:18:16) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license()" for more information.
>>> import ast
>>> ast.dump(ast.parse("True").body[0].value)
'Constant(value=True, kind=None)'
```

On py39+:

```pycon
Python 3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import ast
>>> ast.dump(ast.parse("True").body[0].value)
'Constant(value=True)'
>>> exit()
```

### (2) The AST for slices in between square brackets changed in py39.

No more `ast.Index`.

On py38:

```pycon
>>> ast.dump(ast.parse("Literal[True, False]").body[0].value.slice)
'Index(value=Tuple(elts=[Constant(value=True, kind=None), Constant(value=False, kind=None)], ctx=Load()))'
```

On py39+:

```pycon
>>> ast.dump(ast.parse("Literal[True, False]").body[0].value.slice)
'Tuple(elts=[Constant(value=True), Constant(value=False)], ctx=Load())'
```

Note that for number (2), we elide the difference between py38 and py39 elsewhere in our codebase by getting rid of the `ast.Index` nodes here using the `LegacyNormalizer` before we visit the AST using the `PyiVisitor`:

https://github.com/PyCQA/flake8-pyi/blob/7933e8462578d5a3920296f6bd98ed6547d4cceb/pyi.py#L2136

That doesn't work for the doctests here, though, since you're creating a new AST tree in the doctests that hasn't been transformed by the `LegacyNormalizer`.

### Fixing the doctests

...Maybe we could just get rid of them? We could fix them, but it might be icky. This function seems simple enough that we can probably do without unit tests for now (which is what the doctests _basically_ are), and just rely on the functional tests.